File: Msys_Programs/MSys_Projects/GoProg/DataStructure/LinkedList/circular_doubly.go
package main

import "fmt"

type node struct {
	element int
	next    *node
	prev    *node
}

type CircularDoubly struct {
	haed *node
	tail *node
	size int
}

func (c *CircularDoubly) len() int {
	return c.size
}

func (c *CircularDoubly) isempty() bool {
	return c.size == 0
}

func (c *CircularDoubly) addfirst(e int) {
	newest := &node{e, nil, nil}
	if c.isempty() {
		c.haed = newest
		c.tail = newest
		newest.next = newest
		newest.prev = newest
	} else {
		newest.next = c.haed
		c.haed.prev = newest
		c.tail.next = newest
		c.haed = newest
	}
	c.size++
}

func (c *CircularDoubly) addlast(e int) {
	newest := &node{e, nil, nil}
	if c.isempty() {
		c.haed = newest
		c.tail = newest
		newest.next = newest
		newest.prev = newest
	} else {
		c.tail.next = newest
		newest.prev = c.tail
		c.tail = newest
		c.tail.next = c.haed
		c.haed.prev = c.tail
	}
	c.size++
}

func (c *CircularDoubly) display() {
	p := c.haed
	i := 0
	for i < c.len() {
		fmt.Print(p.element, "->")
		p = p.next
		i++
	}
	fmt.Println()
}

func (c *CircularDoubly) addany(e, position int) {
	if position == 1 {
		c.addfirst(e)
	} else if position >= c.len() {
		c.addlast(e)
	} else {
		newest := &node{e, nil, nil}
		p := c.haed
		i := 1
		for i < position-1 {
			p = p.next
			i++
		}
		newest.prev = p
		newest.next = p.next
		p.next.prev = newest
		p.next = newest
		c.size++
	}

}

func (c *CircularDoubly) removefirst() {
	if c.isempty() {
		fmt.Println("List is empty")
		return
	}
	e := c.haed.element
	c.haed = c.haed.next
	c.haed.prev = c.tail
	c.size--
	fmt.Println("Removed element: ", e)
}

func (c *CircularDoubly) removelast() {
	if c.isempty() {
		fmt.Println("List is empty")
		return
	}
	e := c.tail.element
	c.tail = c.tail.prev
	c.tail.next = c.haed
	c.size--
	// return
	fmt.Println("Removed element: ", e)

}

func (c *CircularDoubly) removeany(position int) {
	if c.isempty() {
		fmt.Println("Array is empty")
		return
	} else if position == 1 {
		c.removefirst()
	} else if position >= c.len() {
		c.removelast()
	} else {
		p := c.haed
		i := 1
		for i < position-1 {
			p = p.next
			i++
		}
		e := p.next.element
		p.next = p.next.next
		p.next.next.prev = p
		c.size--
		fmt.Println(e)
	}
}

func (c *CircularDoubly) displayrev() {
	p := c.tail
	i := 0
	for i < c.len() {
		fmt.Print(p.element, "->")
		p = p.prev
		i++
	}
	fmt.Println()
}

func main() {
	l := &CircularDoubly{}
	l.addfirst(21)
	l.addfirst(22)
	l.addfirst(23)
	l.addfirst(24)
	l.display()
	l.displayrev()
	fmt.Println("Length :", l.len())
	fmt.Println("-----------------------------------------------------------------")
	l.addlast(30)
	l.addlast(31)
	l.display()
	l.displayrev()
	fmt.Println("Length :", l.len())
	fmt.Println("-----------------------------------------------------------------")
	l.addany(50, 5)
	l.display()
	l.displayrev()
	fmt.Println("Length :", l.len())
	fmt.Println("-----------------------------------------------------------------")
	l.addany(55, 8)
	l.display()
	l.displayrev()
	fmt.Println("Length :", l.len())
	fmt.Println("-----------------------------------------------------------------")
	// l.removefirst()
	// l.display()
	// l.displayrev()
	// fmt.Println("Length :", l.len())
	// fmt.Println("-----------------------------------------------------------------")
	// l.removelast()
	// l.display()
	// l.displayrev()
	// fmt.Println("Length :", l.len())
	// fmt.Println("-----------------------------------------------------------------")
	// l.removeany(2)
	// l.display()
	// l.displayrev()
	// fmt.Println("Length :", l.len())
	// fmt.Println("-----------------------------------------------------------------")

}

File: Msys_Programs/MSys_Projects/PythonProg/Python_DSA/LinkedList/circular_doubly.py
class _Node:
    __slots__ = "_element", "_next", "_prev"

    def __init__(self, element, next, prev):
        self._element = element
        self._next = next
        self._prev = prev

class CircularDoubly:

    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def __len__(self):
        return self._size

    def isempty(self):
        return self._size == 0


    def adddfirst(self, e):
        newest = _Node(e, None, None)
        if self.isempty():
            newest._next = newest
            self._head = newest
            self._tail = newest
        else:
            newest._next = self._head
            self._head._prev = newest
            self._tail._next = newest
            self._head = newest
        self._size +=1

    def addlast(self, e):
        newest = _Node(e, None, None)
        if self.isempty():
            newest._next = newest
            self._head = newest
            self._tail = newest
        else:
            self._tail._next = newest
            newest._prev = self._tail
            self._tail = newest
            self._tail._next = self._head
        self._size +=1

    def addany(self, e, position):
        newest = _Node(e, None, None)
        p = self._head
        i = 1
        while i < position -1:
            p = p._next
            i+=1
        newest._next = p._next
        p._next._prev = newest
        p._next = newest
        newest._prev = p
        self._size += 1

    def removefirst(self):
        if self.isempty():
            print("list is empty")
            return
        e = self._head._element
        self._head =self._head._next
        self._head._prev = self._tail
        self._size -=1
        return e

    def removelast(self):
        if self.isempty():
            print("list is empty")
            return
        e = self._tail._element
        self._tail = self._tail._prev
        self._tail._next = self._head
        self._size -= 1
        return e

    def removeany(self, position):
        if self.isempty():
            print("list is empty")
            return
        p = self._head
        i = 1
        while i < position-1:
            p=p._next
            i+=1
        e=p._next._element
        p._next = p._next._next
        p._next._prev = p
        self._size -=1
        return e 

    def display(self):
        p = self._head
        i = 0
        while i < len(self):
            print(p._element,end="->")
            p = p._next
            i+=1
        print()

    def displayrev(self):
        p = self._tail
        i = 0
        while i < len(self):
            print(p._element,end="->")
            p = p._prev
            i+=1
        print()
    
c = CircularDoubly()
c.adddfirst(25)
c.adddfirst(26)
c.adddfirst(27)
c.adddfirst(28)
c.display()
c.displayrev()
print("----------------------------->")
c.addlast(30)
c.addlast(31)
c.addlast(32)
c.addlast(33)
c.display()
c.displayrev()
print("----------------------------->")
c.addany(50, 5)
c.display()
c.displayrev()
print("----------------------------->")
c.removefirst()
c.display()
c.displayrev()
print("----------------------------->")
c.removelast()
c.display()
c.displayrev()
print("----------------------------->")
c.removeany(2)
c.display()
c.displayrev()